### ISSUE TITLE
`.try_collect()` for `Iterator<Option>`

### ISSUE THREAD

**Issue by amab8901** [rust-itertools/itertools#951]
`.try_collect()` is for turning `Iterator<Result<T>>` into `Result<Iterator<T>>`. Can you create an equivalent method that turns `Iterator<Option<T>>` into `Option<Iterator<T>>`? Or perhaps implement it as a generic, as part of `try_collect()`?

The use case is for creating programs that never crash at runtime (where error conditions are passed to the logs, and Result converted to Option, and program is continuing without interruption).

**Comment by Philippe-Cholet** [comment]
`try_collect` does not give `Result<Iterator>` but `Result<Vec<...>>` (or another container than `Vec`).

But in #844 we discussed fallible iterators and we might generalize (amongst other things) `try_collect` (which conflict with nightly Rust) to `maybe_collect` that would solve your use case.

**Comment by amab8901** [comment]
you're right, I missed the Vec part. Here is an example code that would be nice if it worked:

```
use itertools::Itertools;

fn main() {
    // assert!(char::from(1_u8).is_ascii());
    let string: String = [1_u8, 20_u8, 200_u8]
        .iter()
        .map(|u_8| {
            let option_char = if (32..127).contains(u_8) || (160..=255).contains(u_8) {
                let character = char::from(*u_8);
                Some(character)
            } else {
                println!("Error: must be ascii character: `{u_8}`"); // actually tracing::error!() but I wanna keep the example simple
                None
            };

            option_char
        })
        .try_collect()?;

    unreachable!();
}
```

**Comment by Philippe-Cholet** [comment]
Our `maybe_collect` would merely be a stable `Iterator::try_collect`, so `maybe_collect` would eventually be deprecated in favor of it.

**Comment by Philippe-Cholet** [comment]
`try_collect()?` is merely a shortcut for `.collect::<Result<_, _>>()?`

So yeah, there is `.collect::<Option<_>>()?` here.
